keps/sig-cli/kustomize-secret-generator-plugins.md

```diff
@@ -0,0 +1,174 @@
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Kustomize wants to generate Kubernetes Secret objects from general key-value pairs where the value is supposed to be a secret. Currently Kustomize only supports reading secret values from local files which raises security concerns about file lifetime and access. Reading secret values from the execution of arbitrary "commands" in a kustomization.yaml file introduces concerns in a world where kustomization files can be used directly from the internet when a user runs `kustomize build`.
+
+This proposal describes obtaining secret values safely using golang [plugins](https://golang.org/pkg/plugin/).
+
+## Motivation
+
+Not having a way to integrate `SecretGenerator` with secret management tools requires hacky workarounds.
+
+### Goals
+
+- Give users a way to intergrate `SecretGenerator` with secret management tools in a safe way.
+
+### Non-Goals
+
+- Kustomize will not handle plugin installation/management.
+
+## Proposal
+
+In the proposed design `SecretGenerator` would load golang plugins from `~/.config/kustomization_plugins`. This would limit the scope of what kustomize can execute to the plugins on the users machine. Also using golang plugins forces a strict interface with well-defined responsibilities.
+
+The current API looks like this:
+
+```
+secretGenerator:
+- name: app-tls
+  files:
+  - secret/tls.cert
+  - secret/tls.key
+  type: "kubernetes.io/tls"
+
+- name: env_file_secret
+  env: env.txt
+  type: Opaque
+
+- name: myJavaServerEnvVars
+  literals:
+  - JAVA_HOME=/opt/java/jdk
+  - JAVA_TOOL_OPTIONS=-agentlib:hprof
+```
+
+The proposed API would look like this:
+
+```secretGenerator:
+- name: env-example
+  goplugins:
+  - name: Env
+    args:
+    - EXAMPLE_ENV
+    - OTHER_EXAMPLE_ENV
+  type: Opaque
+```
+
+The implementation would look something like this:
+
+```
+func keyValuesFromPlugins(ldr ifc.Loader, plugins []types.Plugin) ([]kv.Pair, error) {
+	var allKvs []kv.Pair
+
+	for _, plugin := range plugins {
+		secretFunc, err := findPlugin(plugin.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		kvs, err := secretFunc(ldr.Root(), plugin.Args)
+		if err != nil {
+			return nil, err
+		}
+
+		allKvs = append(allKvs, kvs...)
+	}
+
+	return allKvs, nil
+}
+
+func findPlugin(name string) (func(string, []string) ([]kv.Pair, error), error) {
+	allPlugins, err := filepath.Glob(os.ExpandEnv("$HOME/.config/kustomization_plugins/*.so"))
+	if err != nil {
+		return nil, fmt.Errorf("error loading plugins")
+	}
+
+	for _, filename := range allPlugins {
+		p, err := plugin.Open(filename)
+		if err != nil {
+			return nil, err
+		}
+
+		symbol, err := p.Lookup(name)
```

Comment:
If we must go with golang plugins, please don't just allow the kustomization.yaml file to invoke any symbol that happens to be exported.

A better way to do this is to have each plugin export a well-known symbol (perhaps `KustomizationSecretGenerators`) that is a map (or array for simplicity) of name => function.  That way only explicitly marked functions will be accessible.

As a bonus, it allows the same plugin .so to be used to provide other sets of entry points (with different function signatures) in the future.  This includes versioning the well-known symbol so we can cleanly evolve the ABI over time.

Reply:
Sounds reasonable.